Only create copy_nop instructions for types for which an encoding exi…

julian-seward1 opened the pull request:
…sts.  Issue #779.

PR #773 detects, at reload time, `copy` instructions that copy a value from stack
slot back to the same stack slot.  It replaces them with `copy_nop` instructions
that have a null encoding (hence producing no code).

For x86_64, `copy_nop` encodings for the types I64, I32, F64 and F32 are
provided.  Unfortunately the code that detects the redundant copy doesn't
check the type of the copied value, hence leaving itself open to the danger of
creating a `copy_nop` instruction cannot be encoded (which is different from
saying it has a null encoding).

This patch:

* Expands the x86_64 set of `copy_nop` encodings to: I64 I32 I16 I8 F64 and F32

* In `visit_inst()` in `reload.rs`, checks the type of the copied value accordingly.

* Adds comments explaining the above.

bjorn3 commented:
It works! Thanks @julian-seward1!

bjorn3 commented:
Could you please add a test like:

```
test compile
target x86_64

function u0:0(i64, i64, i64) system_v {
    sig0 = () system_v
    sig1 = (i16) system_v
    fn1 = u0:94 sig0
    fn2 = u0:95 sig1

ebb0(v0: i64, v1: i64, v2: i64):
    v3 = iconst.i16 0
    jump ebb1(v3)

ebb1(v4: i16):
    call fn1()
    call fn2(v4)
    jump ebb1(v4)
}
```

julian-seward1 commented:
@bjorn3: what does this test check?  What's the rationale?

bjorn3 commented:
It is a reduced reproduction for #779.

bnjbvr commented on cranelift-codegen/meta-python/isa/x86/encodings.py:
```diff
@@ -342,9 +342,15 @@ def enc_i32_i64_ld_st(inst, w_bit, recipe, *args, **kwargs):
 X86_64.enc(base.copy_nop.i64, r.stacknull, 0)
 X86_64.enc(base.copy_nop.i32, r.stacknull, 0)
+X86_64.enc(base.copy_nop.i16, r.stacknull, 0)
+X86_64.enc(base.copy_nop.i8,  r.stacknull, 0)
```
What about boolean types? Do we need to add an encoding for b1 and b8?

julian-seward1 replied:
From the point of view of "will CL assert/fail", no, we don't need them, because the transformation done by reload.rs is now limited to types satisfying `is_int` or `is_float`, and all targets (x86_64, x86_32, rv32, rv64) now have encodings for those types.

From the point of view of "are we missing optimisations", maybe.  But it seems unlikely to me, assuming that most boolean values are consumed by an "if" as soon as they are created.  On that basis, I omitted them.  If profiling later shows evidence that we need these cases, then we can add them.  I haven't seen any need for them in the profiling I've done so far.

bnjbvr commented on cranelift-codegen/src/regalloc/reload.rs:
```diff
@@ -222,15 +222,28 @@ impl<'a> Context<'a> {
-                    // Convert the instruction into a `copy_nop`.
+                    // Convert the instruction into a `copy_nop`.  The
+                    // assertion will fail if there is no encoding for
+                    // `copy_nop` at the required type.
```
I'd remove this comment and instead I'd put a text comment in the assert like `debug_assert!(ok, "copy_nop encoding missing for this type");`

julian-seward1 commented:
New revision; addresses all review comments.  @bnjbvr, in the loops that create the instruction encodings, I used `base.copy_nop.bind(ty)` to create a version of the instruction of type `ty`.  This seems to work but is mere cargo-culting from examples above.  Do you if it is correct?
